source/portable/mbedtls/core_pkcs11_mbedtls.c

```diff
@@ -4095,101 +4266,63 @@ CK_DECLARE_FUNCTION( CK_RV, C_VerifyInit )( CK_SESSION_HANDLE hSession,
         }
     }
 
-    /* Check that a public key was retrieved. */
-    if( xResult == CKR_OK )
-    {
-        /* See explanation in prvCheckValidSessionAndModule for this exception. */
-        /* coverity[misra_c_2012_rule_10_5_violation] */
-        if( xIsPrivate != ( CK_BBOOL ) CK_FALSE )
-        {
-            LogError( ( "Failed to initialize verify operation. Verify "
-                        "operation attempted with private key." ) );
-            xResult = CKR_KEY_TYPE_INCONSISTENT;
-        }
-    }
-
+    /* Retrieve key value from storage. */
     if( xResult == CKR_OK )
     {
         if( 0 == mbedtls_mutex_lock( &pxSession->xVerifyMutex ) )
         {
-            if( ( pxSession->xVerifyKeyHandle == CK_INVALID_HANDLE ) || ( pxSession->xVerifyKeyHandle != hKey ) )
+            switch( pMechanism->mechanism )
             {
-                pxSession->xVerifyKeyHandle = CK_INVALID_HANDLE;
-                mbedtls_pk_free( &pxSession->xVerifyKey );
-                mbedtls_pk_init( &pxSession->xVerifyKey );
-                lMbedTLSResult = mbedtls_pk_parse_public_key( &pxSession->xVerifyKey, pucKeyData, ulKeyDataLength );
+                case CKM_RSA_X_509:
+                case CKM_ECDSA:
 
-                if( 0 != lMbedTLSResult )
-                {
-                    lMbedTLSResult = mbedtls_pk_parse_key( &pxSession->xVerifyKey, pucKeyData, ulKeyDataLength, NULL, 0 );
-
-                    if( 0 != lMbedTLSResult )
+                    /* See explanation in prvCheckValidSessionAndModule for this exception. */
+                    /* coverity[misra_c_2012_rule_10_5_violation] */
+                    if( xIsPrivate != ( CK_BBOOL ) CK_FALSE )
```

Comment:
It seems unnecessary to pass that as a parameter to `prvVerifyInitECRSAKeys`, also the key being public can be assumed in `prvVerifyInitECRSAKeys`. It adds extra complexity to `prvVerifyInitECRSAKeys` to do this.

Reply:
From looking at the API docs, my understanding is that "private" from a PKCS11 perspective means that the key is not exportable. For instance, you could have a private key that is exportable from a secure element and it should set *pIsPrivate = CK_TRUE.

Is this correct?

Reply:
This API is not part of the PKCS #11. It is in `core_pkcs11_pal.h` which is the corePKCS11 PAL, and abstracts non-volatile storage. 

This is just an indication from the PAL to treat the object as a private or public key. Private key objects can use the following attributes to indicate whether they can be exported.

| Attribute | Type | Description |
| -- | -- | -- |
|CKA_EXTRACTABLE | CK_BBOOL | CK_TRUE if key is extractable and can be   wrapped|
|CKA_ALWAYS_SENSITIVE| CK_BBOOL | CK_TRUE if key has always had the   CKA_SENSITIVE attribute set to CK_TRUE |
| CKA_NEVER_EXTRACTABLE | CK_BBOOL | CK_TRUE if key has never had the   CKA_EXTRACTABLE attribute set to CK_TRUE |

This implementation does not respect these attributes however, and just always treats private keys as sensitive and public keys as ex-tractable.

Reply:
So would it be valid for an implementation of PKCS11_PAL_GetObjectValue to report a private key as exportable? 

Alternatively, could a public key be labeled not exportable?

Reply:
If an implementation of the PAL were to do this, the qualification test suite would fail, as the `core_pkcs11_mbedtls.c` will return failures, as it expects, eg. for the PAL to indicate the key is private for a signing operation, and vice versa for verify.